examples/llm_stack/pinecone_module.py

```diff
@@ -0,0 +1,66 @@
+from types import ModuleType
+
+import numpy as np
+import pinecone
+
+
+def client_vector_db(vector_db_config: dict) -> ModuleType:
+    """Instantiate Pinecone client using Environment and API key"""
+    pinecone.init(**vector_db_config)
+    return pinecone
+
+
+def index_name(class_name: str) -> str:
+    """Convert class_name to valid pinecone index_name"""
+    return class_name.lower()
+
+
+def initialize_vector_db_indices(
+    client_vector_db: ModuleType,
+    index_name: str,
+    embedding_dimension: int,
+    embedding_metric: str,
+) -> None:
+    """Initialize Pinecone by creating the index"""
+
+    # do not overwrite if class exists
+    if index_name not in client_vector_db.list_indexes():
+        client_vector_db.create_index(
+            index_name, dimension=embedding_dimension, metric=embedding_metric
+        )
+
+
+def reset_vector_db(client_vector_db: ModuleType) -> None:
+    """Delete the entire index and the data stored"""
+    for idx in client_vector_db.list_indexes():
+        client_vector_db.delete_index(idx)
+
+
+def metadata(embedding_service: str, model_name: str) -> dict:
+    """Create metadata dictionary"""
+    return dict(embedding_service=embedding_service, model_name=model_name)
+
+
+def data_objects(
+    ids: list[str], titles: list[str], embeddings: list[np.ndarray], metadata: dict
+) -> list[tuple]:
+    """Create valid pinecone objects (index, vector, metadata) tuples for upsert"""
+    assert len(ids) == len(titles) == len(embeddings)
+    properties = [dict(title=title, **metadata) for title in titles]
+    embeddings = [x.tolist() for x in embeddings]
+    return list(zip(ids, embeddings, properties))
+
+
+def push_to_vector_db(
+    client_vector_db: ModuleType,
+    index_name: str,
+    data_objects: list[tuple],
+    batch_size: int = 100,
+) -> None:
+    """Upsert objects to Pinecone index"""
+    pinecone_index = pinecone.Index(index_name)
+
+    for i in range(0, len(data_objects), batch_size):
+        i_end = min(i + batch_size, len(data_objects))
+
+        pinecone_index.upsert(vectors=data_objects[i:i_end])
```

Comment:
ditto, anything we could capture and return?

Reply:
Added!